biscuit/src/Auth/Biscuit/Datalog/Parser.hs

```diff
@@ -338,38 +340,38 @@ checkParser = do
   cKind <- l $ choice [ One <$ chunk "check if"
                       , All <$ chunk "check all"
                       ]
-  cQueries <- queryParser
+  cQueries <- queryParser <?> "datalog query"
   pure Check{..}
 
 policyParser :: Parser (Policy' 'Repr 'WithSlices)
 policyParser = do
   policy <- l $ choice [ Allow <$ chunk "allow if"
                        , Deny  <$ chunk "deny if"
                        ]
-  (policy, ) <$> queryParser
+  (policy, ) <$> queryParser <?> "datalog query"
 
 blockElementParser :: Parser (BlockElement' 'Repr 'WithSlices)
 blockElementParser = choice
-  [ BlockRule    <$> try (ruleParser <* C.char ';')
-  , BlockFact    <$> try (factParser <* C.char ';')
-  , BlockCheck   <$> try (checkParser <* C.char ';')
+  [ BlockCheck   <$> checkParser <* C.char ';' <?> "check"
+  , BlockRule    <$> ruleParser <* C.char ';' <?> "rule"
+  , BlockFact    <$> factParser <* C.char ';' <?> "fact"
   ]
 
 authorizerElementParser :: Parser (AuthorizerElement' 'Repr 'WithSlices)
 authorizerElementParser = choice
-  [ AuthorizerPolicy  <$> try (policyParser <* C.char ';')
+  [ AuthorizerPolicy  <$> policyParser <* C.char ';' <?> "policy"
```

Comment:
policyParser itself includes a `<?>`, this will "chain" both messages, is it intended?

Reply:
i'm not sure policy parser includes one

Reply:
Line 351, but that doesn't cover the whole parser. You can ignore this question.